### PR TITLE
Support CREATE OR REPLACE in table materialization

### LINKED ISSUE #371: Support `CREATE OR REPLACE`
### Describe the feature

Hey everyone!

I'm delighted to see `CREATE OR REPLACE` support in Trino: https://github.com/trinodb/trino/issues/13180

This will both:

- History of the table will be preserved in the Iceberg metadata, making it more transparent to the user what happened to the table.
- Make the execution of the pipeline faster, since fewer statements are being issued

### Describe alternatives you've considered

I would love to work on this, but my main question is how to determine if this is supported. I checked, but didn't find any minimal version to enable certain features.

### Who will benefit?

Everyone using Iceberg in Trino 🚀 

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by damian3031**
resolves #371 

Support CREATE OR REPLACE in table materialization.
Other materializations will be handled in separate PRs.

New behaviour is enabled by adding `on_table_exists='replace'` config.

### CODE REVIEW

File: tests/functional/adapter/materialization/test_on_table_exists.py [hunk 1]
```diff
@@ -25,19 +41,40 @@ def project_config_update(self):
...
+        assert len(results) == 1
+        # test tests
+        results = run_dbt(["test"], expect_pass=True)
+        assert len(results) == 3
+
+        # check if the data was loaded correctly
+        check_relations_equal(project.adapter, ["seed", "materialization"])
+
+
+@pytest.mark.iceberg
```
Comment (hovaesco) [inline tests/functional/adapter/materialization/test_on_table_exists.py:63]: delta not supported yet? 

cc @findinpath

File: dbt/include/trino/macros/materializations/table.sql [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {% materialization table, adapter = 'trino' %}
   {%- set on_table_exists = config.get('on_table_exists', 'rename') -%}
-  {% if on_table_exists not in ['rename', 'drop'] %}
+  {% if on_table_exists not in ['rename', 'drop', 'replace'] %}
```
Comment (hovaesco) [inline dbt/include/trino/macros/materializations/table.sql:3]: do you plan to add it to snapshot as well as a part of this PR?

**Review by hovaesco — COMMENTED** [review]
Looks good, this would require docs update.

File: dbt/include/trino/macros/materializations/table.sql (hunk 2 above)
Comment (damian3031) [reply dbt/include/trino/macros/materializations/table.sql:3]: I'll add it in separate PR

File: tests/functional/adapter/materialization/test_on_table_exists.py (hunk 1 above)
Comment (damian3031) [reply tests/functional/adapter/materialization/test_on_table_exists.py:63]: AFAIK it is currently supported only on iceberg
https://github.com/trinodb/trino/issues/13180#issuecomment-1828830623

**Comment by damian3031** [comment]
Latest versions of SEP and Galaxy don't support CREATE OR REPLACE yet (that's why it failed in CI). Therefore, we probably need to wait for this support before merging this PR.

**Comment by wrb2** [comment]
It's a bit confusing, but
1. Trino does support it for Iceberg but not for Delta Lake.
2. SEP does support it for Delta Lake but not for Iceberg.
3. Galaxy documentation is silent on this topic, so I guess it does not support it?

File: tests/functional/adapter/materialization/test_on_table_exists.py (hunk 1 above)
Comment (findinpath) [reply tests/functional/adapter/materialization/test_on_table_exists.py:63]: > delta not supported yet?

`CREATE OR REPLACE` for Delta will be supported as of this week in Galaxy.

OSS see https://github.com/trinodb/trino/pull/19991

Comment (findinpath) [reply tests/functional/adapter/materialization/test_on_table_exists.py:63]: `CREATE OR REPLACE` is live on Galaxy and can be used for testing both Iceberg & Delta Lake tables.

**Comment by wrb2** [comment]
I'm just wondering, does the create or replace drops old data or does the table need vacuum eventually? It has to keep it around at least for a while, otherwise it would break readers.

If it needs vacuum, it might be worth documenting recommendation to run it in a posthook or something, so that the created or replaced tables don't grow out of bounds.

(Unless Trino runs vacuum by itself sometimes.)

**Review by hovaesco — APPROVED** [review]

**Pull request merged**